### PR TITLE
New Building UI - Phase I

### PR DESCRIPTION
**Pull request by Mehuge**
This PR implements the first phase of the building UI, enough to replace the existing building buildblueprints and buildblocks UIs.  Tested on C.U.B.E and Hatchery and working in both.

Please comment, criticise and provide feedback

PR relies upon https://github.com/CUModSquad/Camelot-Unchained-Client-Library/pull/24 and https://github.com/CUModSquad/Camelot-Unchained-Client-Library/pull/25

### CODE REVIEW

File: building/package.json [hunk 1]
```diff
@@ -2,22 +2,22 @@
   "private": true,
   "name": "cu-ui-building",
   "version": "0.1.0",
-  "description": "Camelot Unchained UI - building",
+  "description": "Camelot Unchained UI - respawn",
```
Comment (Mehuge) [inline building/package.json]: building not respawn !fix

File: building/src/index.html [hunk 2]
```diff
@@ -4,13 +4,12 @@
 <!DOCTYPE html>
 <html lang="en">
 <head>
-  <meta charset="UTF-8">
-  <title>Building Mode</title>
-  <link rel="stylesheet" href="../../interface-lib/camelot-unchained/camelot-unchained.css">
-  <link rel="stylesheet" href="css/building.css">
+    <meta charset="UTF-8">
+    <title>Building Blocks &amp; Blueprints</title>
```
Comment (Mehuge) [inline building/src/index.html]: Use and rataher than & because debugger does not display it properly !fix

File: building/src/index.html [hunk 3]
```diff
@@ -4,13 +4,12 @@
...
+    <link rel="stylesheet" href="css/building.css">
 </head>
 <body>
-  <div id="cse-ui-building">
+  <div id="building" class="background"></div>
```
Comment (Mehuge) [inline building/src/index.html]: Should stick to the cse-ui- prefixed ID really for consistency.  !fix

File: building/src/sass/blocks.scss [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          display: block;
+        }
+      }
+      &.shape-icon {
+        -webkit-filter: grayscale(100%);
+        filter: grayscale(100%);
+      }
+    }
+  }
+}
```
Comment (Mehuge) [inline building/src/sass/blocks.scss]: !fix

File: building/src/sass/blueprints.scss [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      background-color: rgba(0,0,0,0.2);
+      outline: none;
+      color: white;
+      width: 170px;    /* 2x button width + 10px spacing */
+      padding: 0 10px;
+      border: 0;
+      box-shadow: inset 0px 0px 5px 2px rgba(0, 0, 0, 0.1);
+    }
+  }
+}
```
Comment (Mehuge) [inline building/src/sass/blueprints.scss]: !fix

File: building/src/sass/building.scss [hunk 6]
```diff
@@ -6,27 +6,46 @@
...
 $cu-styles-root: '../../../interface-lib/camelot-unchained/styles/';
 @import "../../node_modules/camelot-unchained/lib/index.scss";
+@import "settings.scss";
 
-html, body {
-  overflow: hidden;
-  margin: 0;
-  padding: 0;
+html {
+    width: 100%;
```
Comment (Mehuge) [inline building/src/sass/building.scss]: indentation !fix

File: building/src/sass/building.scss [hunk 7]
```diff
@@ -6,27 +6,46 @@
...
+  -webkit-user-select: none;
 }
+.allow-select {
+  -webkit-user-select: all;
+}
+.allow-select-text {
+  -webkit-user-select: text;
+}
+
+#building {
```
Comment (Mehuge) [inline building/src/sass/building.scss]: #cse-ui-building !fix

File: building/src/sass/common.scss [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    width: 5px;
+  }
+}
+
+.loading {
+  width: 100%;
+  margin-top: 64px;
+  color: white;
+  text-align: center;
+}
```
Comment (Mehuge) [inline building/src/sass/common.scss]: !fix

File: building/src/ts/blocks/BlockFilter.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import blocks from '../stores/BlockStore';
```
Comment (Mehuge) [inline building/src/ts/blocks/BlockFilter.tsx]: notused !fix

File: building/src/ts/blocks/BlockFilter.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import { store, BuildingState } from '../stores/Building';
```
Comment (Mehuge) [inline building/src/ts/blocks/BlockFilter.tsx]: store notused !fix

File: building/src/ts/blocks/BlockSection.tsx [hunk 11]
```diff
@@ -0,0 +1,59 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import blocks from '../stores/BlockStore';
```
Comment (Mehuge) [inline building/src/ts/blocks/BlockSection.tsx]: notused !fix

File: building/src/ts/blocks/Blocks.tsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        store.dispatch({ type: 'CLEAR_FILTER' } as any);
+        e.preventDefault();
+        e.stopPropagation();
+        break;
+    }
+  }
+
+  render() {
+    const state: BuildingState = this.props.state;
+    const filter: FilterState = this.props.state.filter;
```
Comment (Mehuge) [inline building/src/ts/blocks/Blocks.tsx]: = state.filter   !fix

File: building/src/ts/blocks/Blocks.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    let statusIcon: string;
+    if (filter.words.length > 0) {
+      statusIcon = 'filtered';
+    } else {
+      statusIcon = 'filter';
+    }
+    let content: JSX.Element[];
+    if (!state.blocks.blocksLoaded) {
+      content = [ <div className="loading"></div> ];
```
Comment (Mehuge) [inline building/src/ts/blocks/Blocks.tsx]: blueprints has a Loading component, it is generic, move it into common, and use it here as well  !fix

File: building/src/ts/blocks/ShapeIcon.tsx [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  onClick = (): void => {
+    store.dispatch({ type: 'SELECT_SHAPE', shape: this.props.shape } as any);
+  }
+
+  render() {
+    const classes: string[] = [ 'block-icon', 'shape-icon', this.props.selected ? 'selected' : undefined ];
+    return (
+      <div className={classes.join(' ')} onClick={this.onClick}>
+        <img src={ "data:image/png;base64," + this.props.icon }/>
+        <ToolTip
```
Comment (Mehuge) [inline building/src/ts/blocks/ShapeIcon.tsx]: This isn't working, tool tips for shapes don't show.  !fix

File: building/src/ts/blocks/ShapeIcon.tsx [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+          id={this.props.id}
+          shape={this.props.shape}
+          icon={this.props.icon}/>
+      </div>
+    );
+  }
+}
+
+export default ShapeIcon;
```
Comment (Mehuge) [inline building/src/ts/blocks/ShapeIcon.tsx]: !fix

File: building/src/ts/blueprints/BlueprintActions.tsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    client.CopyBlueprint();
+  }
+  paste = (e: React.MouseEvent): void => {
+    client.PasteBlueprint();
+  }
+  save = (e: React.MouseEvent): void => {
+    store.dispatch({ type: 'SAVE_BLUEPRINT_MODE', save: true } as any);
+  }
+  delete = (e: React.MouseEvent): void => {
+    cuAPI.DeleteLocalBlueprint(this.props.blueprint.name);    // TODO: Add to client interface
```
Comment (Mehuge) [inline building/src/ts/blueprints/BlueprintActions.tsx]: TODO !fix

File: building/src/ts/blueprints/BlueprintActions.tsx [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    // TODO: Remove selected blueprint from store
```
Comment (Mehuge) [inline building/src/ts/blueprints/BlueprintActions.tsx]: remove comment !fix

File: building/src/ts/blueprints/BlueprintFilter.tsx [hunk 18]
```diff
@@ -0,0 +1,40 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import { client } from 'camelot-unchained';
```
Comment (Mehuge) [inline building/src/ts/blueprints/BlueprintFilter.tsx]: notused !fix

File: building/src/ts/blueprints/BlueprintIcon.tsx [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  select = (): void => {
+    this.props.select(this.props.blueprint);
+  }
+
+  render() {
+    return (
+      <div className={'blueprint-icon' + (this.props.selected ? ' selected' : '')}
+           onClick={this.select}>
+        { this.props.blueprint.icon ?
+            <img src={'data:image/png;base64,' + this.props.blueprint.icon}/>
```
Comment (Mehuge) [inline building/src/ts/blueprints/BlueprintIcon.tsx]: indentation !fix

File: building/src/ts/blueprints/Blueprints.tsx [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+class Blueprints extends React.Component<BlueprintsProps, BlueprintsState> {
+  public name: string = 'Blueprints';
+
+  constructor(props: BlueprintsProps) {
+    super(props);
+  }
+
+  select = (blueprint: Blueprint): void => {
+    cuAPI.SelectBlueprint(blueprint.id);      // TODO: client interface bugged use cuAPI
```
Comment (Mehuge) [inline building/src/ts/blueprints/Blueprints.tsx]: TODO !fix

File: building/src/ts/blueprints/NameInput.tsx [hunk 21]
```diff
@@ -0,0 +1,68 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import { client } from 'camelot-unchained';
```
Comment (Mehuge) [inline building/src/ts/blueprints/NameInput.tsx]: notused !fix

File: building/src/ts/blueprints/SaveBlueprint.tsx [hunk 22]
```diff
@@ -0,0 +1,61 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import { client } from 'camelot-unchained';
```
Comment (Mehuge) [inline building/src/ts/blueprints/SaveBlueprint.tsx]: notused !fix

File: building/src/ts/blueprints/SaveBlueprint.tsx [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+  public name: string = 'SaveBlueprint';
+
+  constructor(props: SaveBlueprintProps) {
+    super(props);
+    this.state = { name: '', ok: false };
+  }
+
+  saveOk = (e: React.MouseEvent): void => {
+    console.log('save blueprint ' + this.state.name);
+    cuAPI.SaveBlueprint(this.state.name);
```
Comment (Mehuge) [inline building/src/ts/blueprints/SaveBlueprint.tsx]: TODO !fix

File: building/src/ts/blueprints/SaveBlueprint.tsx [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+class SaveBlueprint extends React.Component<SaveBlueprintProps, SaveBlueprintState> {
+  public name: string = 'SaveBlueprint';
+
+  constructor(props: SaveBlueprintProps) {
+    super(props);
+    this.state = { name: '', ok: false };
+  }
+
+  saveOk = (e: React.MouseEvent): void => {
+    console.log('save blueprint ' + this.state.name);
```
Comment (Mehuge) [inline building/src/ts/blueprints/SaveBlueprint.tsx]: remove console.log

File: building/src/ts/common/Button.tsx [hunk 25]
```diff
@@ -0,0 +1,45 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import { client } from 'camelot-unchained';
```
Comment (Mehuge) [inline building/src/ts/common/Button.tsx]: notused !fix

File: building/src/ts/common/InputOwnership.ts [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+  else if (e.type === "click" && (e.target as HTMLElement).nodeName === 'BUTTON') {
+    client.ReleaseInputOwnership();
+    if (timer) {  // we released it already
+      clearTimeout(timer);
+      timer = undefined;
+    }
+  }
+}
+
+export default inputOwnership;
```
Comment (Mehuge) [inline building/src/ts/common/InputOwnership.ts]: !fix

File: building/src/ts/dock/Dock.tsx [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import { client } from 'camelot-unchained';
+import * as React from 'react';
+import DockTab from './DockTab';
+import Blocks from '../blocks/Blocks';
+import Blueprints from '../blueprints/Blueprints';
+import BuildingToggle from './BuildingToggle';
+import { BuildingState } from '../stores/Building';
```
Comment (Mehuge) [inline building/src/ts/dock/Dock.tsx]: combine this and next line !fix

File: building/src/ts/dock/Dock.tsx [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+
+class Dock extends React.Component<DockProps, DockState> {
+  public name: string = 'Dock';
+  private tabs: string[] = ['BLOCKS', 'BLUEPRINTS' /* , 'FAVORITES' */];
+
+  constructor(props: DockProps) {
+    super(props);
+  }
+
+  componentDidMount = (): void => {
```
Comment (Mehuge) [inline building/src/ts/dock/Dock.tsx]: notused !fix

File: building/src/ts/dock/Dock.tsx [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+    const classNames: string[] = [ "dock", state.ui.edge ];
+    let content: JSX.Element;
+    if (state.ui.expanded) {
+      arrows = '>>';
+      classNames.push('open');
+      switch(state.ui.tab) {
+        case 'BLUEPRINTS':
+          content = <Blueprints state={this.props.state}/>;
+          break;
+        case 'FAVORITES':
```
Comment (Mehuge) [inline building/src/ts/dock/Dock.tsx]: notused !fix

File: building/src/ts/dock/DockTab.tsx [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+
+export interface DockTabProps {
+  id: string;
+  selected: boolean;
+  select: (tab: string) => void;
+};
+export interface DockTabState {
+};
+
+declare const cuAPI: any;
```
Comment (Mehuge) [inline building/src/ts/dock/DockTab.tsx]: notused !fix

File: building/src/ts/dock/Dock.tsx [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+import { BuildingState } from '../stores/Building';
+import { store } from '../stores/Building';
+
+export interface DockProps {
+  state: BuildingState
+}
+export interface DockState {
+}
+
+declare const cuAPI: any;
```
Comment (Mehuge) [inline building/src/ts/dock/Dock.tsx]: notused !fix

File: building/src/ts/common/Button.tsx [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+
+export interface ButtonProps {
+  onClick?: (e: React.MouseEvent) => void;
+  children?: any;
+  disabled?: boolean;
+};
+export interface ButtonState {
+};
+
+declare const cuAPI: any;
```
Comment (Mehuge) [inline building/src/ts/common/Button.tsx]: notused !fix

File: building/src/ts/main.tsx [hunk 33]
```diff
@@ -4,10 +4,47 @@
...
+
+  // monitor building mode
+  // BUG? doesn't get called with current building mode!
+  client.OnBuildingModeChanged((buildingMode: boolean) => {
+    console.log('building mode ' + buildingMode);
+    store.dispatch({ type: 'SET_BUILDING_MODE', mode: buildingMode } as any);
+  });
+
+  // Turn building mode off, except for C.U.B.E
+  cuAPI.SetBuildingMode(client.patchResourceChannel === 27 ? 1 : 0);
```
Comment (Mehuge) [inline building/src/ts/main.tsx]: TODO !fix

File: building/src/ts/main.tsx [hunk 34]
```diff
@@ -4,10 +4,47 @@
...
  */
 
+// Design Notes: https://docs.google.com/document/d/1H_rSZLFEORrQ_HtJ3JeD12zewzwW5Px33wuN_7n9bDE/edit#
+
 import * as React from 'react';
-import * as ReactDom from 'react-dom';
-import {events} from 'camelot-unchained';
+import * as ReactDOM from 'react-dom';
+import { client, events } from 'camelot-unchained';
+import { Provider } from 'react-redux';
```
Comment (Mehuge) [inline building/src/ts/main.tsx]: notused !fix

File: building/src/ts/main.tsx [hunk 35]
```diff
@@ -4,10 +4,47 @@
...
+import { store } from './stores/Building';
+import blocks from './stores/BlockStore';
+import blueprints from './stores/BlueprintStore';
```
Comment (Mehuge) [inline building/src/ts/main.tsx]: notused !fix

File: building/src/ts/main.tsx [hunk 36]
```diff
@@ -4,10 +4,47 @@
...
+const render = (): void => {
+  ReactDOM.render(
+    <Dock state={store.getState()}/>,
+    document.getElementById('building')
+  );
+};
+
+events.on('init', () => {
+
+  alert('attach debugger');
```
Comment (Mehuge) [inline building/src/ts/main.tsx]: remove !fix

File: building/src/ts/main.tsx [hunk 37]
```diff
@@ -4,10 +4,47 @@
...
+
+  // load block lists
+  blocks.load();
+
+  // monitor building mode
+  // BUG? doesn't get called with current building mode!
+  client.OnBuildingModeChanged((buildingMode: boolean) => {
+    console.log('building mode ' + buildingMode);
```
Comment (Mehuge) [inline building/src/ts/main.tsx]: remove !fix

File: building/src/ts/stores/BlockStore.ts [hunk 38]
```diff
@@ -0,0 +1,153 @@
...
+
+export class BlockStore {
+  blockList: Block[] = [];
+  shapes: ShapeBlock[] = [];
+  types: string[] = [];
+
+  private addBlock = (id: number, icon: string): void => {
+    this.blockList[id] = { id: id, icon: icon, shapes: [], types: [], shape: "", type: "" };
+    // client.RequestBlockTags(id); --- Bugged! (wrong interface)
+    cuAPI.RequestBlockTags(id);
```
Comment (Mehuge) [inline building/src/ts/stores/BlockStore.ts]: TODO !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 39]
```diff
@@ -0,0 +1,211 @@
...
+        this.blueprints[i] = blueprint;
+        return;
+      }
+    }
+    this.blueprints.push(blueprint);
+  }
+
+  private loadIcon = (blueprint: Blueprint): void => {
+    this.updateBlueprint(blueprint);
+    const XHR: XMLHttpRequest = new XMLHttpRequest();
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: move into RestAPI !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 40]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  private addBlueprint = (id: number, name: string): void => {
+    this.loadQ.push({ id: id, name: name, icon: undefined });
+    this.processQ();
+  }
+
+  private downloadBlueprints = (charId: string): void => {
+    // TODO: Move into RestAPI
+    const url: string = 'https://' + cuAPI.webAPIHost + ':4443/api/blueprint?characterID=' + charId + '&loginToken=' + cuAPI.loginToken;
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: Move to RestAPI !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 41]
```diff
@@ -0,0 +1,211 @@
...
+    console.log(url);
+    const XHR: XMLHttpRequest = new XMLHttpRequest();
+    XHR.open('GET', url);
+    XHR.onreadystatechange = (e: ProgressEvent): void => {
+      if (XHR.readyState === 4) {
+        if (XHR.status === 200) {
+          const blueprints: any[] = JSON.parse(XHR.responseText).blueprints;
+          for (var i = 0; i < blueprints.length; i++) {
+            // TODO: Use client API
+            cuAPI.ReceiveBlueprintFromServer(blueprints[i].name, blueprints[i].cellData, blueprints[i]._id);
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: TODO !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 42]
```diff
@@ -0,0 +1,211 @@
...
+    };
+    XHR.onerror = (): void => {
+      console.log('failed to download bluepritns');
+    };
+    XHR.send();
+  }
+
+  private uploadBlueprint = (charId: string, name: string, data: any): void => {
+    // TODO: Move into RestAPI
+    const url: string = 'https://' + cuAPI.webAPIHost + ':4443'
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: Move into RestAPI !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 43]
```diff
@@ -0,0 +1,211 @@
...
+    XHR.open('GET', url);
+    XHR.setRequestHeader('Content-Type', 'application/octet-stream; charset=ascii');
+    XHR.onreadystatechange = (e: ProgressEvent): void => {
+      if (XHR.readyState === 4) {
+        if (XHR.status === 200) {
+          const response: string = XHR.responseText;
+          if (response) {
+            const data: any = JSON.parse(response);
+            if (data.name && data.cellData && data._id) {
+              cuAPI.ReceiveBlueprintFromServer(data.name, data.cellData, data._id);
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: TODO !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 44]
```diff
@@ -0,0 +1,211 @@
...
+    };
+    XHR.onerror = (): void => {
+      console.log('failed to upload blueprint');
+    };
+    XHR.send();
+  }
+
+  private listenBlueprints = (): void => {
+    // TODO: BUG! Missing from client interface
+    cuAPI.OnDownloadBlueprints((charId: string) => {
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: TODO !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 45]
```diff
@@ -0,0 +1,211 @@
...
+    });
+    client.OnNewBlueprint((index: number, name: string) => {
+      console.log('on add blueprint ' + index + ' : ' + name);
+      this.addBlueprint(index, name);
+    });
+    client.OnCopyBlueprint(() => {
+      console.log('on copy blueprint');     // TODO - what's this for?
+      store.dispatch({ type: 'COPY_BLUEPRINT', when: Date.now() } as any);
+    });
+    cuAPI.OnUploadBlueprint((charId: string, name: string, data: any) => {
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 46]
```diff
@@ -0,0 +1,211 @@
...
+      this.load();
+    }
+  }
+  // loads Blueprint details from the server
+  public load = (): void => {
+    store.dispatch({ type: 'LOAD_BLUEPRINTS' });
+    this.blueprints = [];
+    this.listenBlueprints();
+    client.RequestBlueprints();
+    cuAPI.DownloadBlueprints();     // TODO: BUG! Missing from client interface
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: TODO !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 47]
```diff
@@ -0,0 +1,211 @@
...
+    });
+    client.OnCopyBlueprint(() => {
+      console.log('on copy blueprint');     // TODO - what's this for?
+      store.dispatch({ type: 'COPY_BLUEPRINT', when: Date.now() } as any);
+    });
+    cuAPI.OnUploadBlueprint((charId: string, name: string, data: any) => {
+      console.log('on upload blueprint');
+      this.uploadBlueprint(charId, name, data);
+    });
+    cuAPI.OnBlueprintSelected(() => {
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: !fix

File: building/src/ts/stores/BlueprintStore.ts [hunk 48]
```diff
@@ -0,0 +1,211 @@
...
+    XHR.onerror = (): void => {
+      console.log('failed to upload blueprint');
+    };
+    XHR.send();
+  }
+
+  private listenBlueprints = (): void => {
+    // TODO: BUG! Missing from client interface
+    cuAPI.OnDownloadBlueprints((charId: string) => {
+      console.log('on download blueprint ' + charId);
```
Comment (Mehuge) [inline building/src/ts/stores/BlueprintStore.ts]: remove console.logs !fix

File: building/src/ts/stores/Building.ts [hunk 49]
```diff
@@ -0,0 +1,187 @@
...
+  blocks: BlocksState;
+  filter: FilterState;
+}
+
+const building = combineReducers({ ui, selection, blocks, filter });
+
+// ==== store ====
+
+export const store = createStore(building);
+export default store;
```
Comment (Mehuge) [inline building/src/ts/stores/Building.ts]: !fix

File: building/.gitignore [hunk 50]
```diff
@@ -7,8 +7,7 @@ npm-debug.log
 /bin
 /obj
 /tmp
-/tmpp
 /dist
 /publish
 /.vs
-/typings
+/src/tsd
```
Comment (codecorsair) [inline building/.gitignore]: we should be ignoring typings still.
src/tsd should not exist in the directory ever, just delete it rather than ignoring it.

**Comment by codecorsair** [comment]
Closing -- this is live in game but making a new design.

**Pull request merged**